### DISCUSSION TITLE
Improvements to shopping list interaction

### DISCUSSION THREAD

**Discussion by irishrain** [TomBursch/kitchenowl#92] [Ideas]
I wanted to discuss #61. I also feel that the usability of the actual shopping list could be improved, but I don't think double taps are the solution.

**First some thoughts about the current implementation.** It's as simple as it gets. Tap something that you put in your shopping cart and it's gone from the list. This makes sure that the list is not cluttered with old stuff that you already put in. If you tapped something by accident you can scroll down to the recently used items to get it back. While it's intuitive and simple, I have observed several issues with this:

- My first issue is that there is a small delay between tapping an item and the item vanishing. I usually put several items that are close together in my cart, and only then take out my phone to remove them from the list. Sometimes the first item gets removed just as I'm about to tap the next one, moving a different item underneath my finger.

- Even if there is no delay, I have to reorient myself after every tap - while my finger is tapping the first item I already look for the next, which then promptly moves when I remove the first item.

- When I tap something wrong by accident, scrolling down to the recent items to get it back takes quite some time.

- Getting items back this way also removes all the things in the "Description" field.

I also have some **ideas how this could be improved**:

- Implement #61, this would make accidentally removing items less likely, but otherwise it would not be a big improvement. And it would make using the app a lot less intuitive, especially for new users. Double taps is not a common usage pattern on mobile.

- Add an "Undo" button. This would make it easier to fix mistakes, and also could restore the "Description". I don't know how to put this in the UI though, and it would not solve my first two issue.

- Strike through things instead of removing them (this is how a lot of other list apps are doing it). Then provide a button to remove all bought items at once. This would solve all my issues, but means an extra step for the user and means the list might be cluttered with already bought items. I also don't know how where to put the remove button in the UI. I'm also unsure how/when to sync this with the server and other users.

- My favorite: Strike through things instead of removing them. Auto-remove all bought items when the app is closed/screen is turned off/user switches to another view. It would solve all my issues, does not need any new UI item, is pretty intuitive, and no extra effort for the user. It could also be combined with the previous proposal, maybe providing a setting in the options to disable the automatic removal. Has the same open questions about how to sync.

That's what I came up with, maybe someone else has an even better idea?

**Comment by jfdid** [comment]
The strikethrough is a fantastic idea.  It's concrete feedback on completing an item, with no guessing.

**Comment by TomBursch** [comment]
Appreciate the discussion!

> Getting items back this way also removes all the things in the "Description" field.

This should not be the case? It adds all the old description information.

I agree, the strikethrough method is a good idea. Although I think it is best combined with a new FAB. This is more intuitive than auto removing items in a way that is not predictable for users (at least in the beginning).

**Reply by irishrain** [reply]
> > Getting items back this way also removes all the things in the "Description" field.
> 
> This should not be the case? It adds all the old description information.

You are right, it does.

**Comment by freerafiki** [comment]
Hi, thanks for the discussion!

I agree with the issues and while I am not convinced the strikethrough is the best solution, it does solve some issues. We also had the issue about the speed of tapping (even by mistake, you remove the object and put it back many times) and the delay in the sync (the other user would actually see things slightly later creating some confusion), but maybe this is not so easy to solve and also not so important (most likely not so many users will use it at the same time). 

As a small contribution to irishrain idea I have also a couple of suggestions:

1) maybe a "save" button - I mean, when you start tapping the items, they became strike-through and a small "save/save_changes" button appears (or was already there in gray, I am not the best at UI design) that the user can click to finish editing. This would mean the feedback is visible while you are editing (strikethrough) and the synchronization problem is solved in an intuitive way (saving with the button).

2) Otherwise it could be the tapping with strikethrough as a standard, and a secondary option to hold down the item (long hold) and then all items become "selectable" (as I believe this is the case for other apps and also to select multiple messages on popular messaging apps) and then use the "save" button only after a person held down for long.

**Comment by Belgadon** [comment]
I really do like the current solution. Its simple and it works (for me).
Removing an item could be a bit more responsive/fast, that's where I would agree on.

My problem with strike-through (and a small "save" button) is that the list does not get smaller with every item you tap to delete.
And that's something I really like; towards the end of shopping you can see only the items that are missing. I don't have to scroll through 15 rows with 3 items each to see the last two missing (= not strike-through) items.

Bringing back accidently removed items is not a big deal imho.

**Comment by TomBursch** [comment]
Please let me know what you think of the new interaction method here 👇🏼

**Reply by irishrain** [reply]
I used the new method a few times now and I'm happy with it. I have my shopping organized in categories, and usually confirm the ticked off items when I move from one category to the next. Thanks for implementing it!

**Reply by lelutin** [reply]
I tried the shopping list with confirmation instead of tap-remove and I much prefer it. The visual feedback is also better, you see what you chose and the button at the bottom shows that you need the added tap to confirm.

Now having the confirmation instead of tap-remove for ingredients while editing a recipe would be nice :)

**Comment by Quintasan** [comment]
Fresh user here: my biggest problem is that I have a ton of entries in the shopping list because I was fiddling around with the mobile application, and it appears to me that there's *no way* to remove them. Am I missing something?

By removing, I mean removing a shopping list entry for them, not disassociating them from the recipe I created.

**Reply by TomBursch** [reply]
I'm not sure if I understand you correctly, but clicking on the item removes it from the shopping list? If you use the confirmation method, a button should appear after you selected an item and allow you to remove all selected items.
